### ISSUE TITLE
Default values for one canvas become global defaults after reload

### ISSUE THREAD

**Issue by Rhys-T** [Hammerspoon/hammerspoon#3309]
If an `hs.canvas` is given default values for certain element properties (in particular `strokeWidth`, `miterLimit`, and `flatness`), then after reloading the configuration, those values can end up being the global defaults used by _all_ `hs.canvas`es. (Only tested on Catalina so far.)

As far as I can tell, what's happening is:
1. The global defaults for [`strokeWidth`](https://github.com/Hammerspoon/hammerspoon/blob/1bd0c184a6b2acbffad27498b9bb15af3e116b8b/extensions/canvas/libcanvas.m#L456-L462), [`miterLimit`](https://github.com/Hammerspoon/hammerspoon/blob/1bd0c184a6b2acbffad27498b9bb15af3e116b8b/extensions/canvas/libcanvas.m#L354-L360), and [`flatness`](https://github.com/Hammerspoon/hammerspoon/blob/1bd0c184a6b2acbffad27498b9bb15af3e116b8b/extensions/canvas/libcanvas.m#L254-L260) are initialized in the `defineLanguageDictionary` function from their counterparts in the `NSBezierPath` class.
2. [`-[HSCanvasView drawRect:]` copies the defaults for the current canvas into `NSBezierPath`'s defaults](https://github.com/Hammerspoon/hammerspoon/blob/1bd0c184a6b2acbffad27498b9bb15af3e116b8b/extensions/canvas/libcanvas.m#L1421-L1423), but doesn't set them back afterward - and they don't seem to be saved/restored by `NSGraphicsContext`.
3. Hammerspoon reloads the config, and reruns `defineLanguageDictionary`, copying the global defaults out of `NSBezierPath` again… except that those are the values left over from whichever canvas had its `drawRect:` executed last.

This might be as simple as restoring `NSBezierPath`'s defaults at the end of `drawRect:`… but only if anything _else_ that changes them also cleans up after itself. I'm not familiar enough with the APIs to know whether or not that's a safe assumption to make.

<details>
<summary>Minimal example</summary>

```lua
-- ~/.hammerspoon/init.lua

c = hs.canvas.new{x=100, y=100, w=100, h=100}
-- c._default.strokeWidth = 5
c:insertElement{
	type = 'rectangle',
	action = 'stroke',
	strokeColor = { red = 1 },
}
c:show()
```

1. Launch Hammerspoon with the above configuration.
2. Uncomment `c._default.strokeWidth = 5`, and reload. The rectangle's lines will get thicker.
3. Comment it back out, and reload again. The lines _should_ go back to their original width, but they don't.

</details>

**Comment by cmsj** [comment]
@asmagill this might be one for you?

**Comment by asmagill** [comment]
I'll try and look into this in the coming week... I suspect that as stated, I assumed restoring the context would be sufficient to return things to the initial state, but either my assumption was wrong or Apple has changed things a bit... in any case, most of my tests at the time involved a lot of crashing during development so I can't recall now whether I tested this specific behavior after just a reload or not...

**Comment by asmagill** [comment]
Thinking on it, (and I'm including this here more for my notes than any request for comment), it may be best to read in the defaults at load time into a static dictionary, then copy them into the working dictionary during initialization... since the loading of the library only happens the first time, the static dictionary would hold the defaults from the application start and wouldn't be updated during subsequent reloads... have to investigate...

**Comment by asmagill** [comment]
Ok, sorry for taking so long to get back to this, but I have had a chance to try out your sample code and I can verify that it *does* work properly for the latest build of Hammerspoon running on an M1 laptop.

I am currently out of town, so I don't have access to my Intel machine, nor to my machine stuck running Catalina, so I can't test on those.

If you are comfortable building your own Hammerspoon locally, try changing [this line](https://github.com/Hammerspoon/hammerspoon/blob/1bd0c184a6b2acbffad27498b9bb15af3e116b8b/extensions/canvas/libcanvas.m#L4028) to: 

~~~objc
     if (!languageDictionary) languageDictionary = defineLanguageDictionary() ;
~~~

And see if that helps -- I will do the same when I return home in roughly a week.

- - -

I'm also hoping to start a fork in a couple of weeks that specifically targets Catalina -- it's been a while since I've had the time to bring my laptop up to the latest code base and I noticed today that the target no longer allows for even running on Catalina, much less building on it, and since I don't intend to replace my iMac (which only supports up to Catalina) any time soon... we'll see how that goes.

**Comment by Rhys-T** [comment]
Yeah, building the dictionary only once does sound like a simpler solution - not sure why I didn't think of that. I'm afraid I haven't gotten around to installing the full Xcode, just the command-line tools, but if I get a chance to get the correct version downloaded and installed, I'll try building it with that patch.
